### ISSUE TITLE
css justify

### ISSUE THREAD

**Issue by carlsummer** [wkhtmltopdf/wkhtmltopdf#3477]
In the case of both Chinese and English, the article can not be aligned；But in the browser can be achieved
2.png is the effect on the browser, 3.png is the effect of export pdf
![Uploading 2.png…]()
![Uploading 3.png…]()

**Comment by PhilterPaper** [comment]
You posted some code for #3012 -- does that address this problem? 

Are the margins the same here, and just the PDF image is shifted a little left? I'm looking at the place (you marked) where one Chinese character is going over the margin, and nothing jumps out at me as an obvious cause -- the line before also has some Latin characters but does not have this problem. Can you double check whether both lines are getting their Latin text from the same character set? I've seen Chinese fonts that offer both fixed pitch and proportional pitch Latin alphabets, and I'm wondering if the line width was calculated on proportional font, but the rendering was done with a fixed pitch (wider "I", for instance) for the Latin alphabet text? Even if the second line's Latin text includes a character or two from the "wrong" Latin alphabet, it still should have calculated the line width correctly (but at least there will be a clue as to what went wrong).

What is the problem with the second line that you marked? Is the period (?) a little to the left of where it should be? Is that by the same amount that the previous line went past the margin? Again, a possible clue.

**Comment by carlsummer** [comment]
1、No, no solution, can only solve the notes do not contain other labels, if you have other html notes, then the other html notes will be counted into.
![_ 0 no 6 igg x v e24rn](https://cloud.githubusercontent.com/assets/16281965/26612573/47bba36a-45e8-11e7-850d-a3289e59e419.png)
![43qn 7 8 7lej l08e2pag](https://cloud.githubusercontent.com/assets/16281965/26612575/47f42366-45e8-11e7-88fb-8ef9ba811664.png)

2、On the page shows the words are aligned, but in the export pdf, if the Chinese characters in English then there will be some words highlighted or indented, as shown in the figure has a word, there is a fit indent. If the paragraphs are Chinese or English then the left and right sides are aligned

3、Is the same font, font size, etc. are the same

4、Yes, the Chinese is fixed pitch, English is the control pitch so I give that code is the English also become fixed pitch processing, but this deal with the word interval significantly larger.I can not put fixed pitch to proportional pitch, proportional pitch to fixed pitch I was adding a blank to deal with

5、In 2015 in the browser is not able to do the Chinese alignment, but now it can

**Comment by PhilterPaper** [comment]
I'm afraid I'm getting more confused by your description. When you say "html note", are you talking about HTML _tags_ such as `<p>` and `</p>`? As to why they're showing up in the text (in the second example), you should check that the editor (or word processor) you're using isn't outputting `<` and `>` as HTML _entities_ `&lt;` and `&gt;`. If it is, they will show up as literal text and not count as HTML tags.

For (2), if I understand you, if the paragraph is purely Chinese text or purely English text (Latin alphabet), it is properly justified. The problem is only when you mix the two alphabets. However, the first example you show _does_ mix the two alphabets yet properly justifies. What is different in that text? It looks like two paragraphs. Are you using spaces (blanks) to indent, or are you using paragraph tags (`<p>`)? Is the only problem in the first example that character spacing is too large?

For (4), by "control pitch" do you mean "proportional pitch"? That is, variable-width characters? When you say "I give the code is the English also become fixed pitch", is that in a word processor (selecting a fixed pitch Latin alphabet)? In other words, are you using a word processor to create the text you're sending to wkHTMLtoPDF? If so, it may be inserting extra blanks for spacing control, that wkHTMLtoPDF doesn't know how to deal with.

If the input text you're using comes from a word processor instead of a simple text editor, all sorts of hidden stuff could be messing up wkHTMLtoPDF's attempt to turn it into PDF.

**Comment by carlsummer** [comment]
No, it does not mean that, u u, i have the source and the steps to pack you, so maybe you can understand
[question-ans.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/1043372/question-ans.zip)

**Comment by carlsummer** [comment]
No, it does not mean that, u u, i have the source and the steps to pack you, so maybe you can understand
[question-ans.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/1043380/question-ans.zip)

**Comment by PhilterPaper** [comment]
I've been looking at the zip file you sent, and cannot see why you're getting the errors you are. Maybe someone else can take a look at it. You say with "error2" that you are not getting the red characters (span tag with style setting color to red) in either the browser or in the PDF (and you get a Qpainter error in the PDF conversion)?

Again, if you're using a word processor rather than a simple text editor to create your text, check that there are no invisible characters in the HTML text. I'm wondering if there's something in there switching text modes (character width, etc.), or inserting odd spaces for justification, that is confusing even browsers (much less wkHTMLtoPDF).

If that's not it, I'm out of ideas for now. Maybe if I get some free time I'll look at it again, but I can't promise anything. Perhaps there's something in that jQuery JS that's messing up its mind -- have you compared it against a fresh copy? Can you narrow it down to which set of JS is causing the problem?
